noise/README.md

```diff
@@ -207,56 +208,95 @@ handshake payload](#the-libp2p-signed-handshake-payload) as a byte string
-decrypted, the message payload has the structure described in [Encrypted
+* In XX-initiated handshakes, the initiator will send its payload in message 3
+  (closing message), whereas the responder will send theirs in message 2 (their
+  only message).
+* In IK-initiated handshakes, the initiator will optimistically send its payload
+  in message 1 (as it satisfies the guarantee). Next, this case bifurcates:
+    * If the responder continues the IK handshake, it will send its payload in
+      message 2. The handshake ends.
+    * If the responder fall backs to `XXfallback`, it will have failed to
+      decrypt the payload in message 1. A retransmission from the initiator with
+      the fresh cryptographic material is necessary. This is performed in
+      message 3.
+
+When decrypted, the signed payload has the structure described in [Encrypted
 Payloads](#encrypted-payloads), consisting of a length-prefixed `body` field
-followed by optional padding. The `body` of the payload contains a serialized
-[protobuf][protobuf] message with the following schema:
+followed by optional padding.
+
+The `body` of the payload contains a serialized [protobuf][protobuf]
+`NoiseHandshakePayload` message with the following schema:
 
 ``` protobuf
 message NoiseHandshakePayload {
-	bytes libp2p_key = 1;
-	bytes noise_static_key_signature = 2;
-    bytes libp2p_data = 3;
-    bytes libp2p_data_signature = 4;
+	bytes identity_key = 1;
+	bytes identity_sig = 2;
+  bytes data         = 3;
 }
 ```
 
-The `libp2p_key` field contains a serialized `PublicKey` message as defined in
-the [peer id spec][peer-id-spec].
+The `identity_key` field contains a serialized `PublicKey` message as defined
+in the [peer id spec][peer-id-spec].
 
-The `noise_static_key_signature` field is produced using the libp2p identity
-private key according to the [signing rules in the peer id
+The `identity_sig` field is produced using the libp2p identity private key
+according to the [signing rules in the peer id
 spec][peer-id-spec-signing-rules]. The data to be signed is the UTF-8 string
 `noise-libp2p-static-key:`, followed by the Noise static public key, encoded
 according to the rules defined in [section 5 of RFC 7748][rfc-7748-sec-5].
 
-The `libp2p_data` field contains the "early data" provided to the Noise module
-when initiating the handshake, if any. The structure of this data is opaque to
-noise-libp2p and is expected to be defined in a future iteration of the
-connection establishment spec.
-
-If `libp2p_data` is non-empty, the `libp2p_data_signature` field MUST contain a
-signature produced with the libp2p identity key. The data to be signed is the
-UTF-8 string `noise-libp2p-early-data:` followed by the contents of the
-`libp2p_data` field.
+The `data` field contains the "early data" provided to the Noise module when
+initiating the handshake, if any. The structure of this data is opaque to
+noise-libp2p and is defined in the connection establishment specs.
```

Comment:
Do I understand correctly that supporting the sending and receiving of this extra `data` is optional (and thus handshake sealing is optional)? Consequently the only use it may have is for data that is transmitted for some optimisation purposes by other libp2p components but that is never critical for interoperability between different implementations, so if one end sends such data that the other discards, that has no consequence other than the recipient who discarded it missing out on a possible optimisation?

Reply:
We're planning for early handshake data to become key in the new connection establishment flow. We will likely be sharing some form of "connection capabilities" in early data. It's being discussed in the scope of multiselect 2.0:

https://github.com/libp2p/specs/pull/227
https://github.com/libp2p/specs/pull/205

In my head, leveraging early data will be highly encouraged, but ultimately optional, as you say. Not leveraging early data means incurring an extra round trip post-handshake.

Regardless, I would make handshake sealing compulsory for simplicity purposes.